### PR TITLE
bpo-38530: Offer suggestions on AttributeError

### PR DESCRIPTION
**Pull request by pablogsal**
https://bugs.python.org/issue38530

### CODE REVIEW

**Comment by pablogsal** [comment]
@serhiy-storchaka Would you like to take a look at this PR when you have some time? :)

File: Include/cpython/pyerrors.h [hunk 1]
```diff
@@ -62,6 +62,12 @@ typedef struct {
     PyObject *value;
 } PyStopIterationObject;
 
+typedef struct {
+    PyException_HEAD
+    PyObject *obj;
+    PyObject *name;
+} PyAttributeErrorObject;
```
Comment (vstinner) [inline Include/cpython/pyerrors.h:69]: https://www.python.org/dev/peps/pep-0473/ was rejected because it was too broad, but this PR adds a single exception, which sounds ok according to the resolution: https://mail.python.org/pipermail/python-dev/2019-March/156692.html

My main worry is the risk of creating "more" and "worse" exception cycles, but Pablo says that it sounds unlikely: https://bugs.python.org/issue38530#msg354975

File: Include/pyerrors.h [hunk 2]
```diff
@@ -221,7 +221,6 @@ PyAPI_FUNC(PyObject *) PyErr_NewExceptionWithDoc(
     const char *name, const char *doc, PyObject *base, PyObject *dict);
 PyAPI_FUNC(void) PyErr_WriteUnraisable(PyObject *);
 
-
```
Comment (vstinner) [inline Include/pyerrors.h]: revert this unrelated change?

File: Objects/exceptions.c [hunk 3]
```diff
@@ -1338,9 +1338,82 @@ SimpleExtendsException(PyExc_NameError, UnboundLocalError,
...
+    if (BaseException_init((PyBaseExceptionObject *)self, args, NULL) == -1) {
+        return -1;
+    }
+
+    PyObject *empty_tuple = PyTuple_New(0);
+    if (!empty_tuple) {
+        return -1;
+    }
+    if (!PyArg_ParseTupleAndKeywords(empty_tuple, kwds, "|$OO:AttributeError", kwlist,
+                                     &name, &obj)) {
```
Comment (vstinner) [inline Objects/exceptions.c:1358]: Please document the keyword-only parameters somewhere around https://docs.python.org/dev/library/exceptions.html#AttributeError

File: Objects/exceptions.c [hunk 4]
```diff
@@ -1338,9 +1338,82 @@ SimpleExtendsException(PyExc_NameError, UnboundLocalError,
...
+{
+    Py_VISIT(self->obj);
+    Py_VISIT(self->name);
+    return BaseException_traverse((PyBaseExceptionObject *)self, visit, arg);
+}
+
+static PyObject *
+AttributeError_str(PyAttributeErrorObject *self)
+{
+    return BaseException_str((PyBaseExceptionObject *)self);
```
Comment (vstinner) [inline Objects/exceptions.c]: Hum. Can't you simply reuse BaseException_str in ComplexExtendsException()?

File: Objects/object.c [hunk 5]
```diff
@@ -896,17 +918,23 @@ PyObject_GetAttr(PyObject *v, PyObject *name)
         return NULL;
     }
     if (tp->tp_getattro != NULL)
```
Comment (vstinner) [inline Objects/object.c]: Please add { ... } braces when you modify code (PEP 7).

File: Objects/object.c [hunk 6]
```diff
@@ -896,17 +918,23 @@ PyObject_GetAttr(PyObject *v, PyObject *name)
...
-        return (*tp->tp_getattro)(v, name);
-    if (tp->tp_getattr != NULL) {
+        result = (*tp->tp_getattro)(v, name);
+    else if (tp->tp_getattr != NULL) {
         const char *name_str = PyUnicode_AsUTF8(name);
         if (name_str == NULL)
             return NULL;
-        return (*tp->tp_getattr)(v, (char *)name_str);
+        result = (*tp->tp_getattr)(v, (char *)name_str);
+    } else {
```
Comment (vstinner) [inline Objects/object.c]: PEP 7:

```suggestion
    }
    else {
```

File: Objects/object.c [hunk 7]
```diff
@@ -884,10 +885,31 @@ _PyObject_SetAttrId(PyObject *v, _Py_Identifier *name, PyObject *w)
...
+    assert(PyErr_Occurred());
+    // Intercept AttributeError exceptions and augment them to offer
+    // suggestions later.
+    if (PyErr_ExceptionMatches(PyExc_AttributeError)){
+        PyObject *type, *value, *traceback;
+        PyErr_Fetch(&type, &value, &traceback);
+        PyErr_NormalizeException(&type, &value, &traceback);
+        if (PyErr_GivenExceptionMatches(value, PyExc_AttributeError) &&
+            (PyObject_SetAttrString(value, "name", name) ||
+             PyObject_SetAttrString(value, "obj", v))) {
```
Comment (vstinner) [inline Objects/object.c]: Please use _Py_IDENTIFIER() API to avoid decoding byte strings from UTF-8 (and intern the Unicode objects) at each call.

File: Objects/object.c [hunk 8]
```diff
@@ -884,10 +885,31 @@ _PyObject_SetAttrId(PyObject *v, _Py_Identifier *name, PyObject *w)
...
+            return 1;
```
Comment (vstinner) [inline Objects/object.c:902]: Is it really useful to report setattr() failed? the callers don't seem to care.

File: Objects/object.c [hunk 9]
```diff
@@ -896,17 +918,23 @@ PyObject_GetAttr(PyObject *v, PyObject *name)
...
+                    tp->tp_name, name);
     }
-    PyErr_Format(PyExc_AttributeError,
-                 "'%.50s' object has no attribute '%U'",
-                 tp->tp_name, name);
-    return NULL;
+
+    if (!result && add_context_to_attribute_error_exception(v, name)) {
+        return NULL;
+    }
```
Comment (vstinner) [inline Objects/object.c]:
```suggestion
    if (result == NULL) {
        add_context_to_attribute_error_exception(v, name);
    }
```

File: Python/suggestions.c [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+    }
+    if (!suggestion) {
+        return NULL;
+    }
+    return PyUnicode_FromFormat("%S\n\nDid you mean: %U?",
+                                oldexceptionvalue, suggestion);
+}
+
+int
+_Py_offer_suggestions_for_attribute_error(PyAttributeErrorObject* exc) {
```
Comment (vstinner) [inline Python/suggestions.c]: I suggest a shorter name: _Py_attr_error_suggestions().

Usually, C API names use CamelCase: _Py_Suggest_AttributeError()?

File: Python/suggestions.c [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+    int return_val = 0;
+
+    PyObject* name = exc->name;
+    PyObject* v = exc->obj;
```
Comment (vstinner) [inline Python/suggestions.c]:
```suggestion
    PyObject* name = exc->name;  // borrowed reference
    PyObject* v = exc->obj;  // borrowed reference
```

I would also suggest to rename "v" to "obj".

File: Python/suggestions.c [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+    if (!suggestion) {
+        return NULL;
+    }
+    return PyUnicode_FromFormat("%S\n\nDid you mean: %U?",
+                                oldexceptionvalue, suggestion);
+}
+
+int
+_Py_offer_suggestions_for_attribute_error(PyAttributeErrorObject* exc) {
+    int return_val = 0;
```
Comment (vstinner) [inline Python/suggestions.c]: i like to consder that there is an error (init to -1) and only set it to success (0) at the very end, just before exit ;-)

Sometimes, I added an "error:" label just to set it to -1.

Well, it's just a personal preferrence, your code is fine, you can keep it ;-)

File: Python/suggestions.c [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+
+    PyObject* name = exc->name;
+    PyObject* v = exc->obj;
+
+    // Aboirt if we don't have an attribute name or we have an invalid one
+    if ((name == NULL) || (v == NULL) || !PyUnicode_CheckExact(name)) {
```
Comment (vstinner) [inline Python/suggestions.c]:
```suggestion
    if (name == NULL || v == NULL || !PyUnicode_CheckExact(name)) {
```

File: Python/suggestions.c [hunk 14]
```diff
@@ -0,0 +1,230 @@
...
+
+    PyObject* old_args = exc->args;
+    PyObject* new_args = PyTuple_Pack(1, newexceptionvalue);
+    Py_DECREF(newexceptionvalue);
+    if (new_args == NULL) {
+        return_val = -1;
+        goto exit;
+    }
+    exc->args = new_args;
+    Py_DECREF(old_args);
```
Comment (vstinner) [inline Python/suggestions.c]: Py_SETREF() may be useful here.

File: Python/suggestions.c [hunk 15]
```diff
@@ -0,0 +1,230 @@
...
+
+exit:
+    Py_DECREF(oldexceptionvalue);
+
+    // Clear any error that may have happened.
+    PyErr_Clear();
```
Comment (vstinner) [inline Python/suggestions.c]: If the function clears unconditonnaly the current exception, I like to add an assertion to ensure that it's not called with an exception raised.

File: Python/pythonrun.c [hunk 16]
```diff
@@ -1080,6 +1081,12 @@ PyErr_Display(PyObject *exception, PyObject *value, PyObject *tb)
     if (file == Py_None) {
         return;
     }
+
+    if (PyErr_GivenExceptionMatches(exception, PyExc_AttributeError) &&
+        _Py_offer_suggestions_for_attribute_error((PyAttributeErrorObject*) value) != 0) {
+             PyErr_Clear();
```
Comment (vstinner) [inline Python/pythonrun.c]: Please move this code to suggestions.c (add an helper function).

File: Python/suggestions.c [hunk 17]
```diff
@@ -0,0 +1,230 @@
+#include "Python.h"
+
+#include "pycore_suggestions.h"
+
+#define MAX_GETATTR_PREDICT_DIST 3
+#define MAX_GETATTR_PREDICT_ITEMS 100
+#define MAX_GETATTR_STRING_SIZE 20
+
+static size_t
```
Comment (vstinner) [inline Python/suggestions.c:10]: Is that a known algorithm? If yes, tell it in a comment.

File: Python/suggestions.c [hunk 18]
```diff
@@ -0,0 +1,230 @@
...
+                      PyObject* oldexceptionvalue)
+{
+    assert(PyList_CheckExact(dir));
+
+    Py_ssize_t dir_size = PyList_GET_SIZE(dir);
+    if (dir_size >= MAX_GETATTR_PREDICT_ITEMS) {
+        return NULL;
+    }
+
+    int suggestion_distance = PyUnicode_GetLength(name);
```
Comment (vstinner) [inline Python/suggestions.c]: GitHub Actions / Windows (x64)
'initializing': conversion from 'Py_ssize_t' to 'int', possible loss of data [D:\a\cpython\cpython\PCbuild\pythoncore.vcxproj]

File: Python/suggestions.c [hunk 19]
```diff
@@ -0,0 +1,230 @@
...
+        if (!suggestion || current_distance < suggestion_distance) {
+            suggestion = item;
+            suggestion_distance = current_distance;
+        }       
+    }
+    if (!suggestion) {
+        return NULL;
+    }
+    return PyUnicode_FromFormat("%S\n\nDid you mean: %U?",
+                                oldexceptionvalue, suggestion);
```
Comment (vstinner) [inline Python/suggestions.c]: Can you write one line per line, for readability? Something like:

```suggestion
    return PyUnicode_FromFormat("%S\n"
    "\n"
    "Did you mean: %U?",
                                oldexceptionvalue, suggestion);
```

File: Python/suggestions.c [hunk 20]
```diff
@@ -0,0 +1,230 @@
...
+    }
+
+    PyObject* old_args = exc->args;
+    PyObject* new_args = PyTuple_Pack(1, newexceptionvalue);
+    Py_DECREF(newexceptionvalue);
+    if (new_args == NULL) {
+        return_val = -1;
+        goto exit;
+    }
+    exc->args = new_args;
```
Comment (vstinner) [inline Python/suggestions.c]: Why not storing suggestions in a new Exception field, rather than replacing existing args? I may break some applications.

**Review by vstinner — COMMENTED** [review]
First review. I like the overall approach (I expect a low overhead when raising AttributeError).

File: Objects/object.c [hunk 21]
```diff
@@ -884,10 +885,31 @@ _PyObject_SetAttrId(PyObject *v, _Py_Identifier *name, PyObject *w)
     return result;
 }
 
+static inline int
+add_context_to_attribute_error_exception(PyObject* v, PyObject* name)
```
Comment (vstinner) [inline Objects/object.c]: I suggest a shorter name: set_attribute_error_attrs(), or set_attribute_error_context().

But exceptions already have a "__context__" attribute, no? "context" may sound confusing ;-)

File: Python/suggestions.c (hunk 20 above)
Comment (pablogsal) [reply Python/suggestions.c]: Because I don't want to touch the reporting facility to inspect said field or allowing users to inspect it manually

Comment (pablogsal) [reply Python/suggestions.c]: Also, this substitution only happens before printing the value because it has been not caught. Nobody should be able to inspect this exception later, no? Notice this only happens in the default exception display, so if someone has a custom handler, they will not see this AFAIK.

Comment (methane) [reply Python/suggestions.c]: Code level comment:
I don't understand why `exc->args = new_args;` Isn't it set by `Py_SETREF`?

Higher level comment:
Why do you replace `exc->args` instead of just printing suggestion after `_PyErr_Display()` or in `_PyErr_Display()`?

Comment (pablogsal) [reply Python/suggestions.c]: > I don't understand why exc->args = new_args; Isn't it set by Py_SETREF?

I forgot to remove that after updated the code after @vstinner suggestions.

> Why do you replace exc->args instead of just printing suggestion after _PyErr_Display() or in _PyErr_Display()?

Actually, this is a very good point, this can make the code cleaner. I will implement this

File: Doc/library/exceptions.rst [hunk 22]
```diff
@@ -149,6 +149,11 @@ The following exceptions are the exceptions that are usually raised.
    assignment fails.  (When an object does not support attribute references or
    attribute assignments at all, :exc:`TypeError` is raised.)
 
+   The :attr:`name` and :attr:`obj` attributes can be set using keyword-only
+   arguments to the constructor. When set they represent the name of the attribute
+   that was attempted to be accessed and the object that was accessed for said
+   attribute, respectively.
+
```
Comment (vstinner) [inline Doc/library/exceptions.rst:156]: Please add a "versionchanged:: 3.10" markup.

File: Include/internal/pycore_suggestions.h [hunk 23]
```diff
@@ -0,0 +1,12 @@
+#include "Python.h"
+
+#ifndef Py_INTERNAL_SUGGESTIONS_H
+#define Py_INTERNAL_SUGGESTIONS_H
+
+#ifndef Py_BUILD_CORE
+#  error "this header requires Py_BUILD_CORE define"
+#endif
+
+PyObject* _Py_Offer_Suggestions(PyObject* exception);
```
Comment (vstinner) [inline Include/internal/pycore_suggestions.h]: You can put it in Include/internal/pycore_pyerrors.h rather than adding an header file just for a single function declaration.

Comment (vstinner) [inline Include/internal/pycore_suggestions.h]:
```suggestion
extern PyObject* _Py_Offer_Suggestions(PyObject* exception);
```

File: Include/internal/pycore_suggestions.h [hunk 24]
```diff
@@ -0,0 +1,12 @@
+#include "Python.h"
+
+#ifndef Py_INTERNAL_SUGGESTIONS_H
+#define Py_INTERNAL_SUGGESTIONS_H
```
Comment (vstinner) [inline Include/internal/pycore_suggestions.h]: If you want to keep the header file, you should add extern "C". Third party C++ code is allowed to use it.

File: Objects/object.c [hunk 25]
```diff
@@ -884,29 +885,60 @@ _PyObject_SetAttrId(PyObject *v, _Py_Identifier *name, PyObject *w)
...
 {
     PyTypeObject *tp = Py_TYPE(v);
+    PyObject* result = NULL;
 
     if (!PyUnicode_Check(name)) {
         PyErr_Format(PyExc_TypeError,
                      "attribute name must be string, not '%.200s'",
                      Py_TYPE(name)->tp_name);
         return NULL;
     }
```
Comment (vstinner) [inline Objects/object.c:918]: nitpick: I like to only declare variables just before they are initialized (here) thanks to C99.

`PyObject* result;`

File: Objects/object.c [hunk 26]
```diff
@@ -884,29 +885,60 @@ _PyObject_SetAttrId(PyObject *v, _Py_Identifier *name, PyObject *w)
...
+        result = (*tp->tp_getattr)(v, (char *)name_str);
     }
-    PyErr_Format(PyExc_AttributeError,
-                 "'%.50s' object has no attribute '%U'",
-                 tp->tp_name, name);
-    return NULL;
+    else {
+        PyErr_Format(PyExc_AttributeError,
+                    "'%.50s' object has no attribute '%U'",
+                    tp->tp_name, name);
```
Comment (vstinner) [inline Objects/object.c:934]: maybe only set `result = NULL;` in this case.

File: Python/pythonrun.c [hunk 27]
```diff
@@ -953,6 +954,14 @@ print_exception(PyObject *f, PyObject *value)
...
         PyErr_Clear();
     }
+    PyObject* suggestions = _Py_Offer_Suggestions(value);
+    if (suggestions) {
+        err = PyFile_WriteString(". ", f);
+        if (err == 0) {
+            err = PyFile_WriteObject(suggestions, f, Py_PRINT_RAW);
+        }
+        Py_DECREF(suggestions);
+    }
```
Comment (vstinner) [inline Python/pythonrun.c:965]: The API is unclear. You should call PyErr_Clear() if it returns NULL? Or does it only return NULL if there is no suggestion?

File: Python/suggestions.c [hunk 28]
```diff
@@ -0,0 +1,199 @@
...
+    }
+
+    PyObject* suggestions = calculate_suggestions(dir, name);
+    Py_DECREF(dir);
+    return suggestions;
+}
+
+
+// Offer suggestions for a given exception. This function does not raise exceptions
+// and returns NULL if no exception was found.
```
Comment (vstinner) [inline Python/suggestions.c]: "returns NULL if no exception was found"

do you mean "no suggestion"?

Maybe document what it returns. It looks like a Python str object.

File: Python/suggestions.c [hunk 29]
```diff
@@ -0,0 +1,199 @@
...
+PyObject* _Py_Offer_Suggestions(PyObject* exception) {
+    assert(!PyErr_Occurred());
+    if (PyErr_GivenExceptionMatches(exception, PyExc_AttributeError)) {
+        return offer_suggestions_for_attribute_error((PyAttributeErrorObject*) exception);
```
Comment (vstinner) [inline Python/suggestions.c]: "This function does not raise exceptions"

Can you please add an assertion to ensure that?

PyUnicode_FromFormat() can fail with MemoryError.

IMO it's perfectly fine to raise an exception. The caller is free to decide how to handle it.

File: Python/pythonrun.c [hunk 30]
```diff
@@ -1080,6 +1089,7 @@ PyErr_Display(PyObject *exception, PyObject *value, PyObject *tb)
     if (file == Py_None) {
         return;
     }
+
```
Comment (vstinner) [inline Python/pythonrun.c]: THANK YOU! :-D (it makes the code more readable)

File: Python/pythonrun.c [hunk 31]
```diff
@@ -953,6 +954,14 @@ print_exception(PyObject *f, PyObject *value)
     if (err < 0) {
         PyErr_Clear();
     }
+    PyObject* suggestions = _Py_Offer_Suggestions(value);
+    if (suggestions) {
```
Comment (vstinner) [inline Python/pythonrun.c:957]: Would you mind to add a comment showing an example of the expected output?

I understood that it's something like: `. Did you mean: (...)?`.

**Comment by vstinner** [comment]
I like the new implementation, it's even less intrusive! I prefer to not modify `AttributeError.args` when it's displayed. It avoids any risk of backward incompatible change.

File: Objects/object.c (hunk 26 above)
Comment (pablogsal) [reply Objects/object.c:934]: This is code from before, I prefer to not change it in case we introduce some unwanted change

Comment (vstinner) [reply Objects/object.c:934]: Oh ok. It was just a suggestion, you can ignore it. it's just a coding style preference ;-)

File: Python/pythonrun.c (hunk 27 above)
Comment (pablogsal) [reply Python/pythonrun.c:965]: If returns NULL if there is no suggestion. And will never raise exceptions (suggestions are optional).

File: Python/pythonrun.c [hunk 32]
```diff
@@ -15,7 +15,7 @@
 #include "pycore_interp.h"        // PyInterpreterState.importlib
 #include "pycore_object.h"        // _PyDebug_PrintTotalRefs()
 #include "pycore_parser.h"        // _PyParser_ASTFromString()
-#include "pycore_pyerrors.h"      // _PyErr_Fetch
+#include "pycore_pyerrors.h"      // _PyErr_Fetch, _Py_Offer_Suggestions
```
Comment (vstinner) [inline Python/pythonrun.c:18]: FYI I added these comments to check if an include is still needed or not. I put a single symbol. Sometimes, I go through the list of includes to check if they are still needed.

**Review by vstinner — APPROVED** [review]
LGTM.
